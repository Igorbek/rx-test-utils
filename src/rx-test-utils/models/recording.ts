import { WatchNotification } from './messages'

export interface TimeOrderCoordinator {
    nextOrder(time: number): number;
}

export interface TimeOrderAccessor {
    now(): [number, number];
}

export interface RecordedNotification<T> {
    time: [number, number];
    notification: WatchNotification<T>;
}

export interface RecordedSubscription<T> {
    subscribed: [number, number];
    unsubscribed?: [number, number];
    notifications: RecordedNotification<T>[];
}

export interface RecordedObservable<T> {
    subscriptions: RecordedSubscription<T>[];
}
