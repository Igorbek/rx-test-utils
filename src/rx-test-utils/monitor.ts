import { Observer } from 'rxjs/Observer'
import { Observable } from 'rxjs/Observable';

export interface IObservableMonitor<T> {
    subscribe(): IObserverMonitor<T>;
}

export interface IObserverMonitor<T> extends Observer<T> {
    unsubscribe(): void;
}

export function monitored<T>(source: Observable<T>, monitor: IObservableMonitor<T>): Observable<T> {
    return new Observable(observer => {
        const observerMonitor = monitor.subscribe();
        const subscription = source.subscribe(
            value => {
                observerMonitor.next(value);
                observer.next(value);
            },
            error => {
                observerMonitor.error(error);
                observer.error(error);
            },
            () => {
                observerMonitor.complete();
                observer.complete();
            }
        );
        return () => {
            observerMonitor.unsubscribe();
            subscription.unsubscribe();
        }
    });
}
