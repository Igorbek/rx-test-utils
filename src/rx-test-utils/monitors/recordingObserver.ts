import { Observer } from 'rxjs/Observer'
import { Observable } from 'rxjs/Observable'
import { Scheduler } from 'rxjs/Scheduler';

import { WatchNotification } from '../models/messages'
import { RecordedObservable, RecordedSubscription, TimeOrderAccessor } from '../models/recording'

export function createRecordingObserver<T>(
    timeOrderAccessor: TimeOrderAccessor
): {
    observer: Observer<Observable<WatchNotification<T>>>;
    recorded: RecordedObservable<T>;
} {
    const recorded: RecordedObservable<T> = {
        subscriptions: []
    };

    const observer: Observer<Observable<WatchNotification<T>>> = {
        next(value) {
            const {
                observer: subscriptionObserver,
                recorded: recordedSubscription
            } = createRecordingSubscriptionObserver<T>(timeOrderAccessor);
            recorded.subscriptions.push(recordedSubscription);
            value.subscribe(subscriptionObserver);
        },
        complete() {
        },
        error(error) {
            throw error;
        }
    };

    return { observer, recorded };
}

export function createRecordingSubscriptionObserver<T>(
    timeOrderAccessor: TimeOrderAccessor
): {
    observer: Observer<WatchNotification<T>>;
    recorded: RecordedSubscription<T>;
} {
    const recorded: RecordedSubscription<T> = {
        notifications: [],
        subscribed: timeOrderAccessor.now()
    };

    const observer: Observer<WatchNotification<T>> = {
        next(notification) {
            const time = timeOrderAccessor.now()
            recorded.notifications.push({
                time,
                notification
            });
            if (notification.kind === 'U') {
                recorded.unsubscribed = time;
            }
        },
        complete() {
        },
        error(error) {
            throw error;
        }
    };

    return { observer, recorded };
}
