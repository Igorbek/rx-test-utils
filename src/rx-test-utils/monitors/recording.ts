import { TestMessage } from "rxjs/testing/TestMessage";
import { IScheduler } from 'rxjs/Scheduler';
import { Observable } from 'rxjs/Observable';
import { Notification } from 'rxjs/Notification';

import { IObservableMonitor, IObserverMonitor } from '../monitor';

interface RecordedMessageBase {
    time: number;
    order: number;
}

interface RecordedNextMessage<T> extends RecordedMessageBase {
    kind: 'N';
    value: T;
}

interface RecordedCompleteMessage extends RecordedMessageBase {
    kind: 'C';
}

interface RecordedErrorMessage extends RecordedMessageBase {
    kind: 'E';
    error: any;
}

type RecordedMessage<T> = RecordedNextMessage<T> | RecordedCompleteMessage | RecordedErrorMessage;

interface RecordedSubscription<T> {
    subscribedTime: number;
    subscribedOrder: number;
    unsubscribedTime?: number;
    unsubscribedOrder?: number;
    messages: RecordedMessage<T>[];
}

interface RecordedObservable<T> {
    subscriptions: RecordedSubscription<T>[];
}

export function createRecordingMonitor<T>(
    scheduler: IScheduler,
    timeOrderCoordinator: TimeOrderCoordinator): { monitor: IObservableMonitor<T>; recorded: RecordedObservable<T>; } {
    const recorded: RecordedObservable<T> = {
        subscriptions: []
    };

    const monitor: IObservableMonitor<T> = {
        subscribe() {
            const subscribedTime = scheduler.now(), subscribedOrder = timeOrderCoordinator.nextOrder(subscribedTime);
            const subscription: RecordedSubscription<T> = {
                subscribedTime,
                subscribedOrder,
                messages: []
            };
            recorded.subscriptions.push(subscription);

            return {
                next(value) {
                    const time = scheduler.now(), order = timeOrderCoordinator.nextOrder(time);
                    subscription.messages.push({
                        time,
                        order,
                        kind: 'N',
                        value
                    });
                },
                complete() {
                    const time = scheduler.now(), order = timeOrderCoordinator.nextOrder(time);
                    subscription.messages.push({
                        time,
                        order,
                        kind: 'C'
                    });
                },
                error(error) {
                    const time = scheduler.now(), order = timeOrderCoordinator.nextOrder(time);
                    subscription.messages.push({
                        time,
                        order,
                        kind: 'E',
                        error
                    });
                },
                unsubscribe() {
                    subscription.unsubscribedTime = scheduler.now();
                    subscription.unsubscribedOrder = timeOrderCoordinator.nextOrder(subscription.unsubscribedTime);
                }
            }
        }
    };

    return { monitor, recorded };
}

export interface TimeOrderCoordinator {
    nextOrder(time: number): number;
}

export function createTimeOrderCoordinator(): TimeOrderCoordinator {
    let lastTime = 0;
    let nextOrder = 0;
    return {
        nextOrder(time) {
            if (time < lastTime)
                throw new Error('Previous time was higher than passed.');
            if (time > lastTime) {
                lastTime = time;
                nextOrder = 0;
            }
            return nextOrder++;
        }
    }
}
