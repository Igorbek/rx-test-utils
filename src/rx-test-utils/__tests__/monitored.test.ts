import 'jest';

import * as Rx from 'rxjs';
import { TestMessage } from "rxjs/testing/TestMessage";
import { IScheduler } from 'rxjs/Scheduler';
import { Observer } from 'rxjs/Observer'
import { Observable } from 'rxjs/Observable';

import { createRecordingMonitor, createTimeOrderCoordinator } from '../monitors/recording';
import { monitored } from '../monitor';

describe('monitor', () => {
    it('it captures data', () => {
        const testScheduler = new Rx.TestScheduler((a, e) => a === e);
        const { monitor, recorded } = createRecordingMonitor(testScheduler, createTimeOrderCoordinator());

        const source = testScheduler.createHotObservable('--1-2-3--4--5-')
        //const observable = monitored(Rx.Observable.interval(10, testScheduler).take(3), monitor);
        const observable = monitored(source.take(4), monitor);

        const subscription = observable.subscribe();
        const subscription2 = observable.skip(1).subscribe();
        
        //testScheduler.schedule(() => { console.log('done') }, 100);
        testScheduler.flush();

        expect(recorded).toMatchSnapshot();
    })
})
