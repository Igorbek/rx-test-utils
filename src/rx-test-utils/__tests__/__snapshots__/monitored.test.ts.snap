// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`monitor it captures data 1`] = `
Object {
  "subscriptions": Array [
    Object {
      "messages": Array [
        Object {
          "kind": "N",
          "order": 0,
          "time": 20,
          "value": "1",
        },
        Object {
          "kind": "N",
          "order": 0,
          "time": 40,
          "value": "2",
        },
        Object {
          "kind": "N",
          "order": 0,
          "time": 60,
          "value": "3",
        },
        Object {
          "kind": "N",
          "order": 0,
          "time": 90,
          "value": "4",
        },
        Object {
          "kind": "C",
          "order": 1,
          "time": 90,
        },
      ],
      "subscribedOrder": 0,
      "subscribedTime": 0,
      "unsubscribedOrder": 2,
      "unsubscribedTime": 90,
    },
    Object {
      "messages": Array [
        Object {
          "kind": "N",
          "order": 1,
          "time": 20,
          "value": "1",
        },
        Object {
          "kind": "N",
          "order": 1,
          "time": 40,
          "value": "2",
        },
        Object {
          "kind": "N",
          "order": 1,
          "time": 60,
          "value": "3",
        },
        Object {
          "kind": "N",
          "order": 3,
          "time": 90,
          "value": "4",
        },
        Object {
          "kind": "C",
          "order": 4,
          "time": 90,
        },
      ],
      "subscribedOrder": 1,
      "subscribedTime": 0,
      "unsubscribedOrder": 5,
      "unsubscribedTime": 90,
    },
  ],
}
`;
